namecache/namecache.go

```diff
@@ -0,0 +1,151 @@
+// Package namecache implements background following (resolution and pinning) of names
+package namecache
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	merkledag "github.com/ipfs/go-ipfs/merkledag"
+	namesys "github.com/ipfs/go-ipfs/namesys"
+	path "github.com/ipfs/go-ipfs/path"
+	pin "github.com/ipfs/go-ipfs/pin"
+	uio "github.com/ipfs/go-ipfs/unixfs/io"
+
+	logging "gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log"
+)
+
+const (
+	followInterval = 60 * time.Minute
+)
+
+var log = logging.Logger("namecache")
+
+// NameCache represents a following cache of names
+type NameCache interface {
+	Follow(name string, pinit bool)
+	Unfollow(name string)
+	ListFollows() []string
+}
+
+type nameCache struct {
+	nsys    namesys.NameSystem
+	pinning pin.Pinner
+	dag     merkledag.DAGService
+
+	ctx     context.Context
+	follows map[string]func()
+	mx      sync.Mutex
+}
+
+func NewNameCache(ctx context.Context, nsys namesys.NameSystem, pinning pin.Pinner, dag merkledag.DAGService) NameCache {
+	return &nameCache{
+		ctx:     ctx,
+		nsys:    nsys,
+		pinning: pinning,
+		dag:     dag,
+		follows: make(map[string]func()),
+	}
+}
+
+// Follow spawns a goroutine that periodically resolves a name
+// and (when pinit is true) pins it in the background
+func (nc *nameCache) Follow(name string, pinit bool) {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if _, ok := nc.follows[name]; ok {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(nc.ctx)
+	go nc.followName(ctx, name, pinit)
+	nc.follows[name] = cancel
+}
+
+// Unfollow cancels a follow
+func (nc *nameCache) Unfollow(name string) {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	cancel, ok := nc.follows[name]
+	if ok {
```

Comment:
Same as above, If user somehow makes a type there would be no indication

Reply:
will do.

Reply:
done.

Reply:
I would invert this check. `if !ok { return err }`

Reply:
ok, prettier this way